### ISSUE TITLE
Cursor in status line after search

### ISSUE THREAD

**Issue by MVXXC** [vim/vim#2530]
Steps to reproduce the bug:
  1) Open a terminal with 57 lines and 70 columns.
  2) Start Vim (8.0) with the following command line.

       $ vim -u NONE "+se is ls=2"

  3) Open a file with the following contents in this order.
  
       27 lines containing 120 periods each
       1 blank line
       1 line containing the word "target"
       10 lines containing 120 periods each

  4) Place the cursor at the beginning of the file (with "gg").
  5) Search for the word "target" (with "/target").

Description of the bug:
  - When interactive search is on, searching for "target" shows unusual
    behavior: Depending on the number of characters typed, everything works as
    expected (for "t" and "tar", and "targe") or nothing is highlighted (for
    "ta", "targ", and "target").
  - After searching for "t", "tar", and "targe" (where interactive search works
    as expected), the cursor is placed in the status line.

**Comment by tom-m** [comment]
Having spent some time debugging this I think that it's not right when `curwin->w_topline` changes (in the given situation) after each entered character of the search pattern. VIM does not move to any other match. The start of the match is still the same position in the file no matter what further character of the `target` search pattern are given. I don't understand how `curwin->w_topline` is calculated. Anyway, backing up and restoring the starting value for the calculation seems to help.

So here is a patch that solves this issue for me:
```
diff --git a/src/ex_getln.c b/src/ex_getln.c
index 9fdfac5a2..e80f4d354 100644
--- a/src/ex_getln.c
+++ b/src/ex_getln.c
@@ -2043,6 +2043,7 @@ cmdline_changed:
 		SET_NO_HLSEARCH(TRUE);
 
 	    validate_cursor();
+	    old_topline = curwin->w_topline;
 	    /* May redraw the status line to show the cursor position. */
 	    if (p_ru && curwin->w_status_height > 0)
 		curwin->w_redr_status = TRUE;
```
I'd be glad to hear what anybody thinks about this or what you see when testing this solution.

Thanks,

Tom

**Comment by brammool** [comment]
Thanks for looking into this.  Setting old_topline is not the right solution.  The problem is that w_empty_rows is not saved and restored.  I'll make a fix.  Writing a test for this is most of the work...

**Comment by brammool** [comment]
Fixed with  Patch 8.0.1767

**Comment by tom-m** [comment]
On Fri, Apr 27, 2018 at 9:29 PM, Bram Moolenaar <notifications@github.com>
wrote:

> Thanks for looking into this. Setting old_topline is not the right
> solution. The problem is that w_empty_rows is not saved and restored. I'll
> make a fix. Writing a test for this is most of the work...
>
>
> ​Bram,

thanks for the explanation and the proper bug-fix.

Tom​
